### PR TITLE
doc: Add `near-network` reference level documentation `Adding new edges to routing tables`

### PR DESCRIPTION
**Pull request by pmnoxx**
This is part of NetworkDocumentation.md it covers:

-  Adding new edges to routing tables.

See https://github.com/near/nearcore/issues/5156

### CODE REVIEW

File: docs/network.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
+# 7. Adding new edges to routing tables
+
+This section covers the process of adding new edges, received from another nodes,
+to the routing table. It consists of several steps covered below.
+
+## 7.1 Step 1
+`PeerManagerActor` receives `RoutingTableSync` message containing list of new `edges` to add.
+This message is them forwarded to `RoutingTableActor`.
```
Comment (bowenwang1996) [inline docs/network.md:8]:
```suggestion
This message is then forwarded to `RoutingTableActor`.
```

File: docs/network.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+Field `accounts` will be covered in a different section.
```
Comment (bowenwang1996) [inline docs/network.md:9]: Could we add the section number?

File: docs/network.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    /// List of known edges from `RoutingTableActor::edges_info`.
+    edges: Vec<Edge>,
+    /// List of known `account_id` to `PeerId` mappings.
+    /// Useful for `send_message_to_account` method, to route message to particular account.
+    accounts: Vec<AnnounceAccount>,
+}
+```
+
+## 7.2 Step 2
+`PeerManagerActor` forwards those edges to `RoutingTableActor` inside
```
Comment (bowenwang1996) [inline docs/network.md:22]: is this sentence complete?

File: docs/network.md [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    pub(crate) source_peer_id: PeerId,
+}
+```
+
+## 7.3 Step 3
+`RoutingTableActor` get the `ValidateEdgeList` message.
+Filters out `edges` that have already been verified, those that are already in `RoutingTableActor::edges_info`.
+
+Then, it updates `edge_verifier_requests_in_progress` to mark that edge verifications are in progress, and edges shouldn't
+be pruned from Routing Table (see section TODO).
```
Comment (bowenwang1996) [inline docs/network.md:49]: where is this TODO section?

File: docs/network.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+Then, after removing already validated edges, the modified message is forwarded to `EdgeValidatorActor`.
+
+## 7.4 Step 4
+`EdgeValidatorActor` goes through list of all edges.
+It checks whenever all edges are valid (their cryptographic signatures match, etc.).
+
+If any edge is not valid peer will be banned.
+
+Edges that are validated are written to concurrent queue `ValidateEdgeList::sender`/
```
Comment (bowenwang1996) [inline docs/network.md:59]:
```suggestion
Edges that are validated are written to a concurrent queue `ValidateEdgeList::sender`
```

Comment (bowenwang1996) [reply docs/network.md:59]: Also please explain what this concurrent queue does.

File: docs/network.md (hunk 2 above)
Comment (pmnoxx) [reply docs/network.md:9]: @bowenwang1996 What do you mean?

Comment (pmnoxx) [reply docs/network.md:9]: Ok, I added TODO, I'm not sure in which section that will be. My plan is to ship documentation in parts. And then go back to it again, and make fixes, add missing sections, at TODOs. etc.

File: docs/network.md (hunk 4 above)
Comment (pmnoxx) [reply docs/network.md:49]: @bowenwang1996 We are in the middle of rewriting this code. Some PR's are pending review. This section will have to be rewritten.

**Comment by pmnoxx** [comment]
@bowenwang1996 I addresses issues about the doc.

**Review by matklad — COMMENTED** [review]
I am concerned a bit with the level of detail in this docs. We basically copy pats of the code on a very low-level of abstraction here: `Arc<Mutex<HashMap<(PeerId, PeerId), u64>>>`. I am relatively sure that such docs would go stale very quickly. 

I suggest to try to keep the level of abstraction higher here. I'd describe the general algorithm, and what concurrency it intrinsically have *without* references to specific types or modules we use for implementaiton, and then, at the end, give a short paragraph explaining, in a coarse grained way, how the general concepts map to actual implementation. 

If we fundamentally **need** the docs here to be on very low-level of abstraction and attached to specific code, I suggest writing the docs as `//!` comments in code. 


On a meta level, documentation is just like code -- every line written is a line spend, not a line produce. It's better to try to write as little docs as possible, while still conveying all the relevant info.

**Comment by bowenwang1996** [comment]
Agree with what @matklad said!

File: docs/network.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
+# 7. Adding new edges to routing tables
+
+This section covers the process of adding new edges, received from another nodes,
+to the routing table. It consists of several steps covered below.
+
+## 7.1 Step 1
+`PeerManagerActor` receives `RoutingTableSync` message containing list of new `edges` to add.
```
Comment (mm-near) [inline docs/network.md:7]: This is the message that is generated by Step 5, right?

File: docs/network.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    /// List of known `account_id` to `PeerId` mappings.
+    /// Useful for `send_message_to_account` method, to route message to particular account.
+    accounts: Vec<AnnounceAccount>,
+}
+```
+
+## 7.2 Step 2
+`PeerManagerActor` forwards those edges to `RoutingTableActor` inside of `ValidateEdgeList` struct.
+```rust
+pub struct ValidateEdgeList {
```
Comment (mm-near) [inline docs/network.md:24]: this is too much detail - that will be hard to keep in sync. Instead try to describe the gist of the struct (if the name is  not enough).

File: docs/network.md [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+    #[cfg(feature = "test_features")]
+    /// Feature to disable edge validation for purpose of testing.
+    pub(crate) adv_disable_edge_signature_verification: bool,
+    /// Peer that may be banned if any of the edges are found to be invalid.
+    pub(crate) source_peer_id: PeerId,
+}
+```
+
+## 7.3 Step 3
+`RoutingTableActor` get the `ValidateEdgeList` message.
```
Comment (mm-near) [inline docs/network.md:45]: s/get/gets

File: docs/network.md (hunk 7 above)
Comment (pmnoxx) [reply docs/network.md:24]: Ok, I made it simpler.

File: docs/network.md (hunk 6 above)
Comment (pmnoxx) [reply docs/network.md:7]: @mm-near Yes. Once we merge those two PRs. I'll have to go through all points together, and do rewrite, to make sure everything fits.

**Comment by pmnoxx** [comment]
> I am concerned a bit with the level of detail in this docs. We basically copy pats of the code on a very low-level of abstraction here: `Arc<Mutex<HashMap<(PeerId, PeerId), u64>>>`. I am relatively sure that such docs would go stale very quickly.
> 
> I suggest to try to keep the level of abstraction higher here. I'd describe the general algorithm, and what concurrency it intrinsically have _without_ references to specific types or modules we use for implementaiton, and then, at the end, give a short paragraph explaining, in a coarse grained way, how the general concepts map to actual implementation.
> 
> If we fundamentally **need** the docs here to be on very low-level of abstraction and attached to specific code, I suggest writing the docs as `//!` comments in code.
> 
> On a meta level, documentation is just like code -- every line written is a line spend, not a line produce. It's better to try to write as little docs as possible, while still conveying all the relevant info.

Yes, I completely agree.

Initially, I was told by @bowenwang1996 to write the documentation in`github.com/near/NEPs`
by following this example at https://github.com/near/NEPs/blob/master/specs/NetworkSpec/Messages.md
Then, I was told to move it here. 

Anyway, I'm glad you also feel that not having `low-level` abstraction is a good idea. That will make writing the documentation much easier.

**Comment by bowenwang1996** [comment]
> Initially, I was told by @bowenwang1996 to write the documentation in`github.com/near/NEPs`
> by following this example at https://github.com/near/NEPs/blob/master/specs/NetworkSpec/Messages.md
> Then, I was told to move it here.

There are some nuances here. In the spec it is important to highlight the important structures we use. After all, the spec should be written such that other people could build an alternative implementation of the protocol just by reading it. That is why sometimes the spec has pseudocode implementation of some functions and Ethereum does something similar https://github.com/ethereum/consensus-specs/tree/dev/specs. The purpose of the documentation here, however, is more for people to understand this *specific* implementation of the protocol and that is why the focus is different.

**Comment by pmnoxx** [comment]
> > Initially, I was told by @bowenwang1996 to write the documentation in`github.com/near/NEPs`
> > by following this example at https://github.com/near/NEPs/blob/master/specs/NetworkSpec/Messages.md
> > Then, I was told to move it here.
> 
> There are some nuances here. In the spec it is important to highlight the important structures we use. After all, the spec should be written such that other people could build an alternative implementation of the protocol just by reading it. That is why sometimes the spec has pseudocode implementation of some functions and Ethereum does something similar https://github.com/ethereum/consensus-specs/tree/dev/specs. The purpose of the documentation here, however, is more for people to understand this _specific_ implementation of the protocol and that is why the focus is different.

I understand. My point is that I wrote it for another purpose, I'll rewrite It to better fit docs.

File: docs/network.md [hunk 9]
```diff
@@ -62,6 +62,62 @@ Responsibilities:
...
+
+Then, it updates `edge_verifier_requests_in_progress` to mark that edge verifications are in progress, and edges shouldn't
+be pruned from Routing Table (see section TODO).
+
+Then, after removing already validated edges, the modified message is forwarded to `EdgeValidatorActor`.
+
+## 7.4 Step 4
+
+`EdgeValidatorActor` goes through list of all edges.
+It checks whenever all edges are valid (their cryptographic signatures match, etc.).
```
Comment (bowenwang1996) [inline docs/network.md:98]:
```suggestion
It checks whether all edges are valid (their cryptographic signatures match, etc.).
```

**Review by bowenwang1996 — APPROVED** [review]

**Pull request merged**